### ISSUE TITLE
DataLoader tests are quite flaky

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#71187]
Our policy is to disable flaky tests if they are affecting trunk health. Dataloader tests have been quite serious offenders in terms of flakiness (see the long list of [disabled test issues](https://github.com/pytorch/pytorch/issues?q=is%3Aopen+is%3Aissue+label%3A%22module%3A+dataloader%22)). I thought it would be good to start a high level conversation about what we can do to improve the reliability of the dataloader tests.

@mruberry suggests that we split out dataloader tests into a separate config so it is more clear when they fail. I also think that module-level retries are appropriate.

@vitaly-fedyunin, @ejguan: as module owners, who is the right POC to do this stuff?

**Comment by ejguan** [comment]
@mruberry 
What would be the best practice to add a separate config? Do you mean separate workflow?

**Comment by VitalyFedyunin** [comment]
I had chance to take a look at some of this tests, lots of them are related to the distributed training. With this said, can we also separate DL related tests to distributed/non-distributed tiers.

**Comment by mruberry** [comment]
> @mruberry What would be the best practice to add a separate config? Do you mean separate workflow?

Yes; so that dataloader test failures don't appear as failures for other jobs and block signal.

Just hit this, for example: https://github.com/pytorch/pytorch/runs/4848563244?check_suite_focus=true.

**Comment by ejguan** [comment]
Note for failure: I dug a little bit on the failures. Most of them come from we just run out of shared memory especially when DataLoader uses persistent workers.

**Comment by ngimel** [comment]
There's another class of failure caused by name collisions (which could happen if either pids are reused, or atomic increment of file name somehow doesn't work
https://pipelines.actions.githubusercontent.com/mBh68xKhi8LyM7tp3vECvYXNFvuV4gyVGgmYCteuEZP9JH92[…]dSM%2FBcoT7ESxtpRIfGr7TQmrmEJ%2BU%2BmcokWIILGI%3D
https://pipelines.actions.githubusercontent.com/mBh68xKhi8LyM7tp3vECvYXNFvuV4gyVGgmYCteuEZP9JH92QN/_apis/pipelines/1/runs/1591409/signedlogcontent/40?urlExpires=2022-01-19T06%3A58%3A02.1085007Z&urlSigningMethod=HMACV1&urlSignature=V1LL6wPhZN6o8z4etKlvw31ft%2Fyygnmz3SFWf%2B56UpI%3D
The error is 
```
RuntimeError: unable to open shared memory object </torch_10485_0> in read-write mode: File exists (17)
```

**Comment by mruberry** [comment]
Another failure: https://github.com/pytorch/pytorch/runs/4892110531?check_suite_focus=true

```
======================================================================
FAIL [1.821s]: test_get_worker_info (__main__.TestDataLoaderPersistentWorkers)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_dataloader.py", line 1402, in test_get_worker_info
    self.assertEqual(p.exitcode, 0)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2158, in assertEqual
    super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
AssertionError: False is not true : Scalars failed to compare as equal! 1 != 0
```

**Comment by mruberry** [comment]
Another failure to open a shared memory object: https://github.com/pytorch/pytorch/runs/4905695066?check_suite_focus=true

```
test_multiprocessing_contexts (__main__.TestDataLoaderPersistentWorkers) ... [W CudaIPCTypes.cpp:15] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]
[W CudaIPCTypes.cpp:15] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]
[W CudaIPCTypes.cpp:15] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]
[W CudaIPCTypes.cpp:15] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]
[W CudaIPCTypes.cpp:15] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]
[W CudaIPCTypes.cpp:15] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/multiprocessing/queues.py", line 245, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/opt/conda/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "/opt/conda/lib/python3.9/site-packages/torch/multiprocessing/reductions.py", line 249, in reduce_tensor
    event_sync_required) = storage._share_cuda_()
  File "/opt/conda/lib/python3.9/site-packages/torch/storage.py", line 622, in _share_cuda_
    return self._storage._share_cuda_(*args, **kwargs)
RuntimeError: unable to open shared memory object </torch_12110_0> in read-write mode: File exists (17)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/multiprocessing/queues.py", line 245, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/opt/conda/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "/opt/conda/lib/python3.9/site-packages/torch/multiprocessing/reductions.py", line 249, in reduce_tensor
    event_sync_required) = storage._share_cuda_()
  File "/opt/conda/lib/python3.9/site-packages/torch/storage.py", line 622, in _share_cuda_
    return self._storage._share_cuda_(*args, **kwargs)
RuntimeError: unable to open shared memory object </torch_12109_0> in read-write mode: File exists (17)
```

**Comment by seemethere** [comment]
> @mruberry What would be the best practice to add a separate config? Do you mean separate workflow?

We really need to add documentation on how to add separate configurations for GHA

**Comment by suo** [comment]
Alright @VitalyFedyunin, @ejguan, what do we want to do here? I would like to make forward progress on this, so we should decide:
- Split it out into separate job?
- Fix the shm ooms and name collisions?
- both?

**Comment by ejguan** [comment]
> Split it out into separate job?

I will look into it next week to see how to achieve so.

> Fix the shm ooms and name collisions?

My PR may only solve part of the problem (unfortunately reverted). But, fixing the name collisions is extremely hard to me as I have never been able to reproduce it.

@seemethere Do you have any suggestion on it since (IIUC) you suggested parallel execution of tests in GHA may cause this issue more flaky?

**Comment by VitalyFedyunin** [comment]
we would prefer to do both, @seemethere do we have any sort of examples to make (configs separation) happen?
meanwhile we are going to look at collisions and ooms

**Comment by ngimel** [comment]
Users also report a similar error re: exising file, which may or may not be related the errors we are seeing in CI https://github.com/pytorch/pytorch/issues/67864#issuecomment-1020741011. Note, the error message reported in the issue is different, but it's because we changed it after 1.10 to report better information.

**Comment by ejguan** [comment]
After https://github.com/pytorch/pytorch/pull/71579 is landed, we potentially are going to solve both problems in DataLoader.
1. Reset connection 
2. Zombie persistent process in DataLoader (the problem that we run out of shared memory) 

we would observe the flakiness of the tests and see if we want to invest more time on the separate workflow.

**Comment by suo** [comment]
Okay, we'll try to keep an eye on things. If there are no additional flaky errors, we can try un-skipping some of tests.

@janeyx99 ideally a flaky test disable system could automate this process.

**Comment by ngimel** [comment]
#72232 should solve name collisions in a brute force way (or rather, make them rare enough so that they are not a problem)
